src/Fmt_ast.ml

```diff
@@ -56,6 +56,15 @@ let rec sugar_or_pat c ({ast= pat} as xpat) =
       @ sugar_or_pat c (sub_pat ~ctx pat2)
   | _ -> [xpat]
 
+let rec protect_token c = function
+  | Typ typ -> (
+    match (typ.ptyp_desc, c) with
+    | Ptyp_arrow (_, _, t2), _ -> protect_token c (Typ t2)
+    | Ptyp_object _, (']' | '}') -> true
+    | _ -> false )
+  | Pld (PTyp t) -> protect_token c (Typ t)
+  | _ -> false
```

Comment:
This function has a similar functionality to `exposed` in `Ast`, but for types instead of expressions. It might make sense to move it there.